ANW-617: Add Human Readable URLs to public interface

avatar382 opened the pull request:
Adding human readable URLs top level entity types in the public interface.

Entity types with human readable URLs:
-Repostitories
-Agents
-Subjects
-Resources
-Digital Objects
-Accessions
-Classifications

Slugs can be viewed and modified in the frontend, in the entity edit form.
Slugs can be set to automatically generate based on name/title or identifier. 
Automatically generated slugs will update themselves when name/title or identifier is updated.
Slugs will automatically de-duplicate and truncate themselves to 50 chars.

Background job added to frontend to automatically create slugs and enable autogeneration for all entities in the instance. This is intended for the first time setup.

avatar382 commented:
I've added the bugfixes and enhancements we talked about, which includes:

-Showing a preview of the public link in the show and edit pages for the entities with slug support
-Disabling the slug field in the form when autogenerate slug is turned on
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This seems to have resolved the strangeness we were seeing with slugs getting de-duped when they actually weren't duplicates
-Tweaks to ensure that the generate slug background job completely resets the slug

avatar382 commented:
Updated with:

- Slug generation for archival objects and digital object components in frontend forms and background job
- Public interface displays slugged links for digital object components and archival objects in entity tree
- Public interface routing correctly handles slugged links for archival objects and digital object components

lmcglohon commented on public/app/views/repositories/_repository.html.erb:
```diff
@@ -1,8 +1,10 @@
-<%= (full ? '<h1>' : '<h2>').html_safe %>
-<% if !full %><a class="record-title" href="<%= app_prefix(result['uri']) %>"><% end %><%= result['name']%>
+
+<%= (full ? '<h1>' : '<h3>').html_safe %>
```
This should be changed to 
```suggestion
<%= (full ? '<h1>' : '<h2>').html_safe %>
```

avatar382 replied:
Changed back in bdbdc92.

lmcglohon commented on backend/spec/factories.rb:
```diff
@@ -712,6 +712,9 @@ def JSONModel(key)
     source  { create(:json_resource).uri }
   end
 
+  factory :generate_slugs_job, class: JSONModel(:generate_slugs_job) do
```
empty factory here, so not needed?

lmcglohon replied:
I think this is fine, then.

lmcglohon commented:
@avatar382 Tests are failing.  Can you rebase to the current master once you have removed files with blank lines as the only changes? Then we can see how the tests do.

avatar382 commented:
Updated branch with review fixes, and a config option to hide slugs from the PUI.

Looking at Travis test failures, they're all in the public tests. I'm not able to get the public tests to pass in my dev environment, in the master branch. @lmcglohon are the public tests passing for you?

lmcglohon commented on backend/app/lib/slug_helpers.rb:
```diff
@@ -0,0 +1,423 @@
+
+    # REPO FOUND - find entity
+    else
+      found_in = nil
```
is this necessary?  Already initialized to nil in line 117.

lmcglohon commented on backend/app/lib/slug_helpers.rb:
```diff
@@ -0,0 +1,423 @@
+  end
+
+
+  # given a slug, return true if slug is used by another entitiy.
```
mispelling for entity in comment

lmcglohon commented on backend/app/lib/slug_helpers.rb:
```diff
@@ -0,0 +1,423 @@
+           do_component_count > 0
+  end
+
+  # dupe_slug is already in use. Recusively find a suffix (e.g., slug_1)
```
mispelling for recursively in comment

lmcglohon commented:
@avatar382 I believe there may be an issue with classifications.  If there is a classification_term for a classification, in either the view or edit mode, the classification will not display correctly. Screens shots attached include browse classifications page, classification with classification_term edit page and classification without classification term edit page.
Browse
<img width="985" alt="browsepage" src="https://user-images.githubusercontent.com/3585151/49464981-37120500-f7b1-11e8-87ba-ac6cee6fe7f6.png">

View or Edit Page for Classification with Classification Term
<img width="1335" alt="withclassificationtermeditpage" src="https://user-images.githubusercontent.com/3585151/49464980-36796e80-f7b1-11e8-82cc-196d2cb6eb68.png">

Edit Page for Classification without Classification Term
<img width="1316" alt="withoutclassificationtermeditpage" src="https://user-images.githubusercontent.com/3585151/49464979-36796e80-f7b1-11e8-992d-72ef18d58e5e.png">

lmcglohon commented:
@avatar382 Can you rebase this branch and renumber the database migrations? I am having trouble getting the URLs to work and am wondering if it is because I did both of locally...

avatar382 commented:
@lmcglohon Added the changes discussed last meeting:

- migration to generate slugs for repos based on repo_code
- hide frontend slug UI elements when AppConfig[:slugs] is set to hide
- remove consecutive, leading and trailing underscores from slugs
- rebased on latest master (12/7)

avatar382 commented:
Rebased with latest master.

avatar382 commented:
Updated based on latest review:

1. slug_helpers.rb:172 - rename methods to have name like *_any_repo:

> Simple method rename, only in two places in the same class, so this one was easy.

1. slug_helpers.rb:407.- use AS.app_prefix in aspace_form_helper.rb, try using AS.app_prefix to generate a full URL

> Looking into this, I realized that AS.app_prefix is not currently a component that's usable in the backend or frontend Ruby code. It's only defined as a JS method in the frontend, and a Ruby module in the public service. So, to keep things simple, I've left it as is. In the future we can look into writing a shared component in common that does this.

1. large_tree.rb: cleanup to see if stuff can be removed -> explain why it needs to be kept
1. trees.rb: cleanup to see if stuff can be removed -> explain why it needs to be kept
clean up console.logs from tree_render.rb

> Removed unnecessary code from backend/app/model/large_tree.rb and  backend/app/model/mixins/trees.rb, and cleaned up comments.

1. slug_helpers.rb: private method = move to backend/app/model/mixins/identifiers.rb

> moved format_multipart_identifier method to identifiers mixin, and made all helper methods private.

1. Why don’t we have an include for archival_object and classification_terms slug JS file?

> I tested this and it was working without the includes, the JS is probably getting included up the chain somewhere. However, I've added the missing include statements for clarity and to prevent an unexpected failure in case the JS is refactored later.

1. Clean out extra ReadOnlyContext in aspace_form_helper.

> Removed extra class.

1. In jobs form, only show EAD_ID copy if generate slugs by ID is enabled.

> Updated view logic to so copy only shows when appropriate.

1. Refactor params handling public/application:75,79 to make it clearer that original params aren’t being wiped out.

> Refactored public/application_controller.rb to use a hash merge instead of passing and modifying the params hash in other methods.

1. Fix public/view_helper indentation.

> Indentation fixed.

1. Why don’t we have a method for classification terms and digital_object_component?

> Looked into this: In the public app, all Classification Term and Digital Object Component links are created by the LargeTree code, so a method for these in the helper is not needed.
